WORKSPACE

```diff
@@ -35,16 +34,10 @@ git_repository(
 # CUDA should be installed on the system locally
 new_local_repository(
     name = "cuda",
-    path = "/usr/local/cuda-11.1/",
+    path = "/usr/local/cuda",
```

Comment:
Does this work if people have CUDA 11.2 installed as their primary cuda? I think we'd accidentally end up linking the wrong library. 

I think in general the way we have approached WORKSPACE is as a declaration of dependencies, and its incumbent on the user to correctly modify the WORKSPACE for their particular environment, all we can do is ensure that what we provides guarantees correct compilation on our conception of the default system environment (so Ubuntu 18.04 with latest CUDA install via debians). Given that latest CUDA is 11.2 and presumably be linked to `/usr/local/cuda`, we should set the specific path to ensure people are using the right version and let them change it if its different.

Reply:
This was particularly a problem when PyTorch was 6 months behind a major version update to CUDA

Reply:
Yes that worked for me - 11.1 path may work for people who had 11.1 installed, but I installed 11.2 as my first CUDA 11 and don't want to go through installing 11.1 and then reinstalling 11.2. 
I understand your considerations about not wanting people to get "successful" build while potentially linking wrong library - but people do expect GitHub projects to build out of the box these days and early build breaks may be discouraging. 
   The way you prefer people to build right now automatically means custom changes to WORKSPACE - not just uncommenting some lines, they actually need to decide where to put downloaded archive and how to refer to it (would have been so much better if there were no login requirements at developer.nvidia.com). Besides first build frustration, it also means content issues with contribution - one has to exclude those manual changes to WORKSPACE before submitting  PR...
   So, perhaps, better solution here would be to remove WORKSPACE from the project and provide WORKSPACE.local and WORKSPACE.archive templates instead, documenting that user should use one of them for their WORKSPACE and implications of using each ?

Reply:
> 11.1 path may work for people who had 11.1 installed, but I installed 11.2 as my first CUDA 11 and don't want to go through installing 11.1 and then reinstalling 11.2.

I think the solution for your particular case to avoid uninstalling an reinstalling is to install the `cuda-libraries-11-1 cuda-libraries-dev-11-1` packages (I think these are the right ones) which can be installed after 11.2 and wont override default cuda but give you the libraries in `/usr/local/cuda-11.1`. IIRC this is what did since like you 11.2 is my original CUDA. Ideally we could make CUDA hermetic too but I have no idea how we could do that. 

> but people do expect GitHub projects to build out of the box these days and early build breaks may be discouraging.

Yeah I get that but I think its something that you cant really control considering how building C++ projects work. We try to be as hermetic as possible because I find the downstream errors of slight library version mismatches or system configuration incompatibilities much more frustrating and difficult to debug than file path issues. We default to archive because I don't think its is too hard to download two tarfiles and put them in a directory and use the `--dist-dir` flag like we put in the documentation and README and be guaranteed they are working with the same libraries we are targeting https://nvidia.github.io/TRTorch/tutorials/installation.html#building-using-cudnn-tensorrt-tarball-distributions
(yeah totally agree if NVIDIA's distribution of cuDNN and TensorRT was more liberal it would be way nicer). I feel if people want to do something non standard or something we can't fully account for, they should explicitly make that choice and so if they hit errors they can refer back to this step when debugging. 

> it also means repeated issues with contributions - one has to exclude those manual changes to WORKSPACE before submitting PR...

Yeah I guess it means you can't do `git add -u`. `git add core/  tests/` is what I have been using. We can maybe add a git hook to catch attempting to commit WORKSPACE changes for you.

> So, perhaps, better solution here would be to remove WORKSPACE from the project and provide WORKSPACE.local and WORKSPACE.archive templates instead, documenting that user should use one of them for their WORKSPACE and implications of using each ?

We could do this and gitignore WORKSPACE @peri044 thoughts? My concern is its another file that needs to be updated every version bump since I'm still not convinced that leaving the path as `/usr/local/cuda` is the right thing to do for most users.

Reply:
Thanks for the ideas! 
I understand your concern about updating stuff in WORKSPACE when we bump version of anything. 
Can that be solved by some kind of WORKSPACE that includes .bzl files some of them has to be user-defined (like WORKSPACE.trt-deps.bzl) ? So that the bulk of WORKSPACE is checked in except for the configurable part that users have to make decision on and maintain themselves (I see two of them: WORKSPACE.libtorch.bzl default/local and WORKSPACE.trt-deps) 
I quickly tried including .bzl in WORKSPACE which by itself worked but then ran into some issues. So just sharing this idea here so Bazel pros can ponder :)

Reply:
> I don't think its is too hard to download two tarfiles and put them in a directory and use the --dist-dir flag like we put in the documentation and README and be guaranteed they are working with the same libraries we are targeting https://nvidia.github.io/TRTorch/tutorials/installation.html#building-using-cudnn-tensorrt-tarball-distributions

I have tried using --dist-dir : not really convenient to type every single time. Is there a way to put it to per-project rc file? 
Another - probably the simplest - way would be to direct users to put the tarballs in dist_dir/x86_64-linux-gnu and have checked in WORKSPACE refer to it - then first time, the build would break with meaningful diagnostic. That's what I do for myself locally - did not find the way to refer to the dir relatively to the project - though I did not try too hard to make it portable.

Reply:
We could do macros for versions, that might work. I would need to look into how that works exactly with WORKSPACE. 

You can add command aliases to the `.bazelrc`file. We could maybe do something like `--archive-deps-x86_64` or something shorter which aliases to `--dist-dir thrid_party/distdir/x86_64-linux-gnu` or add it as something that is added by default but people who use local would need to disable that. 

We could also explore something like this https://stackoverflow.com/questions/58937356/bazel-relative-local-path-as-url-in-http-archive

Reply:
That certainly calls for another PR. Reverted local/archive changes.